Inactive subscriber calculation tweaks [MAILPOET-5410]

johnolek opened the pull request:
## Description

This PR makes some significant changes to the way engagement scores are calculated and updated.
- Sending a newsletter to a subscriber marks them as being due for an engagement score recalculation the next time the cron runs.
- if a subscriber currently has an engagement score but hasn't received a newsletter in the last year, their engagement score will be reset to "unknown" instead of simply remaining the same.
- We're no longer counting machine opens as opens for the engagement score calculation.
- All of the individual stats on a subscriber's statistics page are now based on data from the last year, not all time. The UI has been updated with text to make this clear.
- The WC revenues stats also now include data just for the last year.
- Individual stats pages now include a new section displaying dates of last engagement of different types (click, open, page view, etc).

## Code review notes

_N/A_

## QA notes

If possible, could you please benchmark a larger send before and after this change? It would be good to know if the additional query is going to have a noticeable affect on performance.

## Linked PRs

_N/A_

## Linked tickets

https://mailpoet.atlassian.net/browse/MAILPOET-5410

## After-merge notes

_N/A_

## Tasks

- [x] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [x] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes

Commit:
Add bulk update engagement score updated at method

MAILPOET-5410

Commit:
Nullify engagement score updated at after sends

This will cause the cron job that recalculates engagement score to pick
up these users the next time it runs. This ensures scores don't get
stale.

MAILPOET-5410

Commit:
Reset engagement score for inactive subscribers

If someone hasn't received at least 3 emails in the last year, we can't
trust that their level of engagement would be the same now as it was
over a year ago.

MAILPOET-5410

Commit:
Make stats consistent with engagement score

This makes SubscriberStatisticsRepository the source of truth for
engagement score data instead of having separate queries.

As part of this change, we will now only be displaying the last year's
worth of data when viewing a report for an individual subscriber.

It also updates engagement score calculation to only count human opens,
not machine opens.

MAILPOET-5410

Commit:
Update woo revenue stats to only look at past year

MAILPOET-5410

Commit:
Add subheading to subscriber stats page

MAILPOET-5410

Commit:
Add engagement summary to subscriber stats

MAILPOET-5410

Commit:
Fix tests

MAILPOET-5410

brezocordero commented on mailpoet/lib/Subscribers/SubscribersRepository.php:
```diff
@@ -304,6 +304,21 @@ public function bulkUpdateLastSendingAt(array $ids, DateTimeInterface $dateTime)
+      ->setParameter('ids', $ids)
+      ->getQuery()
+      ->execute();
+    return count($ids);
```
What do you think about removing this return value?
It does not look like we use it. And I know we return this same count of ids in other update functions (might have done it myself) but it does not provide any information about how many ids where actually upated.

johnolek replied:
Very good point! I'll remove this.

brezocordero commented on mailpoet/assets/js/src/subscribers/stats.tsx:
```diff
@@ -4,10 +4,12 @@ import { MailPoet } from 'mailpoet';
 import { useGlobalContextValue } from 'context';
 
-import { StatsHeading as Heading } from './stats/heading';
+import { Heading } from 'common/typography/heading/heading';
```
It looks like the imports were simplified [here](https://github.com/mailpoet/mailpoet/commit/c4ce4fd10b8b55b3c1f536ba01a04ee11caaa7b1), in case you want to do the same.

johnolek replied:
Thanks for pointing this out! Updated in 59f5ff3d9a54304d01757fe8bf3736079db25c20

brezocordero commented on mailpoet/lib/Cron/Workers/SendingQueue/SendingQueue.php:
```diff
@@ -274,6 +274,8 @@ private function processSending(SendingTask $queue, int $timer): void {
           $now = Carbon::createFromTimestamp((int)current_time('timestamp'));
           $this->subscribersRepository->bulkUpdateLastSendingAt($foundSubscribersIds, $now);
+          // We're nullifying this value so these subscribers' engagement score will be recalculated the next time the cron runs
+          $this->subscribersRepository->bulkUpdateEngagementScoreUpdatedAt($foundSubscribersIds, null);
```
You have probably thought about this but why not modify the bulkUpdateLastSendingAt and make this update in the same query?

johnolek replied:
I did think it would be nice to execute both of these updates in the same query, but I wanted to keep the methods separate in case we ever wanted to use them separately in other contexts, and to keep them easier/clearer to test.

My guess was that this wouldn't cause much of a performance hit, but it's probably worthwhile to do everything we can to minimize the query count during the sending process.

I decided to wrap the updates in a transaction in 0783c6ec0db9e47298f503957ff7b682b6ee326, which should address this issue for anyone using InnoDB. Does this seem like a reasonable approach?

costasovo replied:
@johnolek My knowledge of InnoDB isn't very profound. So just to confirm I understand. Using the transaction still means DB needs to run two queries, but the expected performance gain comes from the data being written only once. Correct?

johnolek replied:
@costasovo All I was thinking was that since InnoDB supports transactions and MyISAM doesn't, if a user is using InnoDB this might result in only a single round trip to the database. I think I may have been mistaken though, and using a transaction doesn't necessarily mean the operations would only require a single round trip.

Do you think it would be worthwhile to create a more generic bulk update method, one that accepts key/value pairs and can bulk update multiple columns in a single query? 

My suspicion is that the code should be ok as is -- processing a sending queue already requires lots of separate DB queries and this one shouldn't be very slow. I might ask if QA can do a before/after test to see if this change affects sending time for very large sends. We can always optimize this later if we discover it's causing performance issues.

brezocordero commented on mailpoet/assets/js/src/subscribers/stats.tsx:
```diff
@@ -87,6 +88,15 @@ export function SubscriberStats(): JSX.Element {
           }}
         />
+        <EngagementSummary
+          lastClick={stats.last_click}
+          lastEngagement={stats.last_engagement}
+          lastOpen={stats.last_open}
+          lastPageView={stats.last_page_view}
+          lastPurchase={stats.last_purchase}
+          lastSending={stats.last_sending}
```
These properties are optional on the StatsType, should we check for existence or make them required?

johnolek replied:
It's ok for them to be `undefined` but I agree that it would be nicer to be more explicit about the fact that these are optional. I've refactored the `EngagementStats` component in 3dceb386e0157b189a2bf61b221eb3c9960e446 to hopefully do a better job of that.

brezocordero reviewed (dismissed):
@johnolek This is great work.
I have some small questions/comments but I have approved the PR because they are not required.
Feel free to assign to QA if you prefer to ignore them.

Commit:
Remove unused and unhelpful return value

MAILPOET-5410

Commit:
Simplify import

MAILPOET-5410

Commit:
Wrap bulk updates in a transaction

MAILPOET-5410

Commit:
Refactor engagement summary

MAILPOET-5410

johnolek commented:
Thanks for the review, @brezocordero ! I think I've addressed all your comments in my last few commits. Since you're AFK for the next couple of weeks I'll ask someone else on the team to take a second look.

johnolek commented:
@costasovo Brezo did an initial review of this PR and left for a couple weeks of AFK before I had a chance to complete the revisions. Would you have time to take a second look?

costasovo reviewed (approved):
Thanks @johnolek for looking into this. This will make the engagement score more useful.

Aschepikov reviewed (approved):
GJ!
